File: 404. Sum of Left Leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // DFS: O(N)time, O(1)space, 2.85%, ok
    int sumOfLeftLeaves(TreeNode* root) {
        int sum = 0;
        dfs(root, sum, false);
        return sum;
    }

    void dfs(TreeNode* root, int& sum, bool isLeft){
        if(!root) return;
        if(!root->left && !root->right && isLeft)
            sum += root->val;
        dfs(root->left, sum, true);
        dfs(root->right, sum, false);
    }

    // DFS without marker: O(N)time, no space, 2.85%, ok
    int sumOfLeftLeaves(TreeNode* root) {
        return dfs(root);
    }

    int dfs(TreeNode* root){
        if(!root) return 0;
        if(root->left && !root->left->left && !root->left->right)
            return root->left->val + dfs(root->right);
        else return dfs(root->left) + dfs(root->right);
    }
};

File: 404. Sum of Left Leaves.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # dfs: O(N) time, O(h) space
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(root, isLeft):
            curr = 0
            if (not root.left) and (not root.right):
                return root.val if isLeft else 0
            if root.left:
                curr += dfs(root.left, True)
            if root.right:
                curr += dfs(root.right, False)
            return curr
        
        return dfs(root, False) if root else 0

    # bfs, queue: O(N) time & space
    def sumOfLeftLeaves(self, root):
        if not root:
            return 0
        queue, ret = [(root, False)], 0
        for node, isLeft in queue:
            if not node.left and not node.right and isLeft:
                ret += node.val
            if node.left:
                queue.append((node.left, True))
            if node.right:
                queue.append((node.right, False))
        return ret
        
            
        